model/jpa/src/main/java/org/keycloak/models/jpa/JpaIdentityProviderStorageProvider.java

```diff
@@ -218,16 +221,27 @@ public IdentityProviderModel getByAlias(String alias) {
     }
 
     @Override
-    public Stream<IdentityProviderModel> getAllStream(Map<String, String> attrs, Integer first, Integer max) {
+    public Stream<IdentityProviderModel> getAllStream(IdentityProviderQuery query, Integer first, Integer max) {
         CriteriaBuilder builder = em.getCriteriaBuilder();
-        CriteriaQuery<IdentityProviderEntity> query = builder.createQuery(IdentityProviderEntity.class);
-        Root<IdentityProviderEntity> idp = query.from(IdentityProviderEntity.class);
+        CriteriaQuery<IdentityProviderEntity> cq = builder.createQuery(IdentityProviderEntity.class);
+        Root<IdentityProviderEntity> idp = cq.from(IdentityProviderEntity.class);
 
         List<Predicate> predicates = new ArrayList<>();
         predicates.add(builder.equal(idp.get("realmId"), getRealm().getId()));
 
-        if (attrs != null) {
-            for (Map.Entry<String, String> entry : attrs.entrySet()) {
+        List<String> includedProviderFactories = null;
+        if (query.getType() != null && query.getType() != IdentityProviderType.ANY) {
+            includedProviderFactories = IdentityProviderTypeUtil.listFactoriesByType(session, query.getType());
+        } else if (query.getCapability() != null) {
```

Comment:
This means that if both `type` and `capability` are filled in the query, the `capability` will be ignored. This is probably OK as there is probably no use-case for search by both type and capability. But maybe it is a bit cleaner to have this checked on the `IdentityProviderQuery` as well? So for example if someone tries to fill:
```
idpQuery.type(CLIENT_AUTHENTICATION).capability(LINK_USERS)
```
it will throw some exception like `unsupported to fill capability when type already filled` . Or at least consider this in javadoc?

Reply:
Ah, ok. I did not looked carefully. Thanks!